### DISCUSSION TITLE
Setting up AVD1001A with pfSense as router

### DISCUSSION THREAD

**Discussion by jsnyder** [brianschrameck/arlo-cam-api#10] [General]
Hi,

So far I have had success with getting some VMC4040Ps and VMC4030Ps online.  I do have the VMC4030Ps going through battery faster than I'd like but I'm thinking I'll just wire those so I can use them with the NVR.

What I am running into though is after pairing the AVD1001 back with my hub (was previously using it as a cloud camera) and then letting it join the network I set up for the cameras (got the PSK & network name) I can't seem to get a stream from it.  I'm wondering if even though I did set up some NAT rules if they're not quite right since this is one of the cameras that depends more on that working.

I have a VLAN set up for the Arlo cameras and they get IPs from that (Unifi APs).  I used the same IP range that was in the examples (172.14.1.0/24). LAN is 192.168.0.0/23.  The base station emulator is on LAN and is at 192.168.0.112. Scrypted is at 192.168.1.107. Then I have these rules:
<img width="1150" alt="image" src="https://github.com/user-attachments/assets/e3309cb7-8eee-447f-80af-165c858f7c66">

<img width="1150" alt="image" src="https://github.com/user-attachments/assets/0f050c7f-a75e-4638-b55e-60f48662abbb">

I guess I'm not entirely sure how this two-piece NAT is working where it knows which camera to send connections back to.

Since this works at all, I'm presuming these rules are somewhat doing what they should, but maybe not.  I also figured if there's something different here this could help folks with other router setups.

It does get snapshots from it but video doesn't seem to work for either TCP or UDP.

When I try playing video in the scrypted WebUI, it has log entries like this:
```
[WebRTC Plugin]: Connection is local network: true 192.168.1.6 {
[WebRTC Plugin]:   ipv4: true,
[WebRTC Plugin]:   type: 'prflx',
[WebRTC Plugin]:   isLocalNetwork: true,
[WebRTC Plugin]:   destinationId: '192.168.1.6'
[WebRTC Plugin]: }
[WebRTC Plugin]: done waiting ice connected
[WebRTC Plugin]: Connection is local network: true 192.168.1.6 {
[WebRTC Plugin]:   ipv4: true,
[WebRTC Plugin]:   type: 'prflx',
[WebRTC Plugin]:   isLocalNetwork: true,
[WebRTC Plugin]:   destinationId: '192.168.1.6'
[WebRTC Plugin]: }
[Rebroadcast Plugin]: Selected stream Stream 1
[Rebroadcast Plugin]: Stream 1 prebuffer session started
[Rebroadcast Plugin]: rtsp outgoing message
[Rebroadcast Plugin]:  OPTIONS rtsp://172.14.1.128/live RTSP/1.0
[Rebroadcast Plugin]: CSeq: 0
[Rebroadcast Plugin]: User-Agent: Scrypted
[Rebroadcast Plugin]: 
[Rebroadcast Plugin]: client error Error: read ECONNRESET
[Rebroadcast Plugin]:     at TCP.onStreamRead (node:internal/stream_base_commons:218:20) {
[Rebroadcast Plugin]:   errno: -104,
[Rebroadcast Plugin]:   code: 'ECONNRESET',
[Rebroadcast Plugin]:   syscall: 'read'
[Rebroadcast Plugin]: }
[Rebroadcast Plugin]: rtsp outgoing message
[Rebroadcast Plugin]:  TEARDOWN rtsp://172.14.1.128/live RTSP/1.0
[Rebroadcast Plugin]: CSeq: 1
[Rebroadcast Plugin]: User-Agent: Scrypted
[Rebroadcast Plugin]: 
[Rebroadcast Plugin]: Stream 1 prebuffer session ended with error r [Error]: Operation Timed Out
[Rebroadcast Plugin]:     at Timeout._onTimeout (/@scrypted/prebuffer-mixin/main.nodejs.js:1:41293)
[Rebroadcast Plugin]:     at listOnTimeout (node:internal/timers:581:17)
[Rebroadcast Plugin]:     at process.processTimers (node:internal/timers:519:7) {
[Rebroadcast Plugin]:   promise: Promise { <pending> }
[Rebroadcast Plugin]: }
[WebRTC Plugin]: Error starting playback for WebRTC track. RPCResultError: Operation Timed Out
[WebRTC Plugin]:     at Timeout._onTimeout (/@scrypted/prebuffer-mixin/main.nodejs.js:1:41293)
[WebRTC Plugin]:     at listOnTimeout (node:internal/timers:581:17)
[WebRTC Plugin]:     at process.processTimers (node:internal/timers:519:7)
[WebRTC Plugin]: thread:main
[WebRTC Plugin]: @scrypted/prebuffer-mixin:host
[WebRTC Plugin]: host:@scrypted/nvr
[WebRTC Plugin]: @scrypted/nvr:host
[WebRTC Plugin]: host:@scrypted/webrtc {
[WebRTC Plugin]:   cause: undefined
[WebRTC Plugin]: }
[WebRTC Plugin]: signalingState closed
[WebRTC Plugin]: connectionState closed
[WebRTC Plugin]: iceConnectionState closed
stopping stream
```

and around the same time server.py prints this:
```
2024-08-02 15:03:05 INFO     192.168.1.107 - - [02/Aug/2024 15:03:05] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:03:16 INFO     192.168.1.107 - - [02/Aug/2024 15:03:16] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:03:27 INFO     192.168.1.107 - - [02/Aug/2024 15:03:27] "POST /device/A2R1987MB5801/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:03:30 INFO     192.168.1.107 - - [02/Aug/2024 15:03:30] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:03:38 INFO     192.168.1.107 - - [02/Aug/2024 15:03:38] "POST /device/A2R1987MB5801/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:03:41 INFO     192.168.1.107 - - [02/Aug/2024 15:03:41] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
Exception in thread Thread-10135:
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/root/arlo-cam-api/server.py", line 58, in run
    msg = self.connection.receive()
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/arlo-cam-api/arlo/socket.py", line 25, in receive
    raise RuntimeError("socket connection closed")
RuntimeError: socket connection closed
Exception in thread Thread-10136:
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/root/arlo-cam-api/server.py", line 58, in run
    msg = self.connection.receive()
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/arlo-cam-api/arlo/socket.py", line 25, in receive
    raise RuntimeError("socket connection closed")
RuntimeError: socket connection closed
2024-08-02 15:04:00 INFO     192.168.1.107 - - [02/Aug/2024 15:04:00] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:04:10 INFO     192.168.1.107 - - [02/Aug/2024 15:04:10] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:04:30 INFO     192.168.1.107 - - [02/Aug/2024 15:04:30] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:04:38 INFO     192.168.1.107 - - [02/Aug/2024 15:04:38] "POST /device/A2R1987MB5801/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:04:40 INFO     192.168.1.107 - - [02/Aug/2024 15:04:40] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:04:48 INFO     192.168.1.107 - - [02/Aug/2024 15:04:48] "POST /device/A2R1987MB5801/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:05:02 INFO     192.168.1.107 - - [02/Aug/2024 15:05:02] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
2024-08-02 15:05:12 INFO     192.168.1.107 - - [02/Aug/2024 15:05:12] "POST /device/A2G308KAA4E19/userstreamactive HTTP/1.1" 200 -
```

Is this an indication that I've got something wrong with the rules or do I need to change something else?  The VMC4040Ps work with TCP, and the VMC4030Ps seem to need to be UDP.

Also, any recommendations for what settings to use with Unifi that the cameras seem to be happy with?  They seem to all be 2.4 GHz, but I didn't really mess with defaults much.

Thanks for maintaining this and in advance for any suggestions.

**Comment by brianschrameck** [comment]
The rules seem right to me, but the behavior you are seeing indicates otherwise and looks similar to what would happen if the camera were to see streaming requests coming from something other than the default Arlo gateway. What is the `ARLO address` for the SNAT translation address? That should be set to the gateway IP: `172.14.1.1`.

Also not sure about the `Interface` for your SNAT rule, shouldn't that be the `ARLO` interface?

**Comment by jsnyder** [comment]
I'm not totally sure what ARLO address, I figured it was the gateway but it could be the subnet? (though in other places when the address is the subnet it is called out as such).

In any case after your suggestions I did play around with it more and it also clicked that what we're doing is having devices reach out to the individual IPs on that subnet and what we want is anything that connects to appear as if it is the gateway so I switched it to take any traffic going to the ARLO subnet and translate the source IP to the gateway:
<img width="1153" alt="image" src="https://github.com/user-attachments/assets/b68a23f6-f185-40ad-a0f2-397eddb52942">

That works and ditto for the other cameras still working.  In general does everything going to that camera have to appear as if it is the gateway? i.e.: RTSP clients as well as what's going through the base station to communicate with the camera?  Or does it just need to be the base station? (I have separate containers & IP for the base station and where scrypted is running).

Also, it looks like it times out after some period of connection, so I presume you can't record from this camera continuously?  Is that timeout and/or cooldown period something that is known?

**Reply by brianschrameck** [reply]
That's a good question, I'm not totally sure. I do seem to remember having issues with the Video Doorbell streaming for longer periods, but I chalked it up to other strange behavior I was seeing with it even with the native Arlo system.

I'm pretty sure the RTSP requests (e.g. Scrypted) need to appear to come from the Arlo gateway. The commands from Arlo API I don't think have to. No harm in including both, though.

**Reply by jsnyder** [reply]
Ok. I'll experiment with it and will add a note if I find some consistency to it. Maybe it's also my unit.  I definitely appreciate giving some more utility to these cameras.

**Reply by jsnyder** [reply]
I'm having it drop off the wifi and then get into a state where it's inaccessible (after a while stops attempting to join WiFi) without doing a reset of the device.  Maybe there's something up with my particular device or it doesn't like something about the WiFi setup.  The other models seem to be doing alright.  I have had it do stuff like this before and then eventually got to a somewhat stable setup. I don't know exactly what fixed it before though :-)
